### PR TITLE
Report GlobalKey duplicates

### LINKED ISSUE #6109: Be better about catching duplicate GlobalKeys
```
15:39:00.293: I/flutter : ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
15:39:00.296: I/flutter : The following assertion was thrown building Flexible(flex: 1):
15:39:00.300: I/flutter : scheduleBuildFor() called for a widget for which a build was already scheduled.
15:39:00.303: I/flutter : The method was called for the following element:
15:39:00.306: I/flutter :   DecoratedBox(dirty; renderObject: RenderDecoratedBox DETACHED)
15:39:00.309: I/flutter : The current dirty list consists of:
15:39:00.312: I/flutter :   [DecoratedBox(dirty; renderObject: RenderDecoratedBox DETACHED), RawGestureDetector(state:
15:39:00.315: I/flutter :   RawGestureDetectorState(192277613; gestures: tap; behavior: opaque)), Text("Suggestion Ichi";
15:39:00.318: I/flutter :   inherit: true; color: Color(0xffffffff); size: 16.0; letterSpacing: 1.92x; height: 1.5x),
15:39:00.321: I/flutter :   Image(image: AssetImage(bundle: null, name: packages/armadillo/res/flutter-mark-square-100.png);
15:39:00.324: I/flutter :   color: Color(0xffffffff); fit: ImageFit.cover; state: _ImageState(623862140; stream:
15:39:00.327: I/flutter :   ImageStream(OneFrameImageStreamCompleter; [100×100] @ 1.0x; 65 listeners); pixels: [100×100] @
15:39:00.330: I/flutter :   1.0x)), Image(image: AssetImage(bundle: null, name: packages/armadillo/res/Armadillo.png); fit:
15:39:00.333: I/flutter :   ImageFit.cover; state: _ImageState(435251538; stream: ImageStream(OneFrameImageStreamCompleter;
15:39:00.336: I/flutter :   [1230×1230] @ 1.0x; 25 listeners); pixels: [1230×1230] @ 1.0x)), DecoratedBox(renderObject:
15:39:00.340: I/flutter :   RenderDecoratedBox), RawGestureDetector(state: RawGestureDetectorState(604374343; gestures: tap;
15:39:00.342: I/flutter :   behavior: opaque)), Text("Suggestion Ni"; inherit: true; color: Color(0xffffffff); size: 16.0;
15:39:00.345: I/flutter :   letterSpacing: 1.92x; height: 1.5x), Image(image: AssetImage(bundle: null, name:
15:39:00.348: I/flutter :   packages/armadillo/res/flutter-mark-square-100.png); color: Color(0xffffffff); fit:
15:39:00.351: I/flutter :   ImageFit.cover; state: _ImageState(460373810; stream: ImageStream(OneFrameImageStreamCompleter;
15:39:00.353: I/flutter :   [100×100] @ 1.0x; 65 listeners); pixels: [100×100] @ 1.0x)), Image(image: AssetImage(bundle: null,
15:39:00.356: I/flutter :   name: packages/armadillo/res/Armadillo.png); fit: ImageFit.cover; state: _ImageState(224197766;
15:39:00.359: I/flutter :   stream: ImageStream(OneFrameImageStreamCompleter; [1230×1230] @ 1.0x; 25 listeners); pixels:
15:39:00.362: I/flutter :   [1230×1230] @ 1.0x))]
15:39:00.364: I/flutter : This usually indicates that a widget was rebuilt outside the build phase (thus marking the element
15:39:00.367: I/flutter : as clean even though it is still in the dirty list). This should not be possible and is probably
15:39:00.370: I/flutter : caused by a bug in the widgets framework. Please report it:
15:39:00.373: I/flutter : https://github.com/flutter/flutter/issues/new
15:39:00.376: I/flutter : To debug this issue, consider setting the debugPrintScheduleBuildForStacks and
15:39:00.379: I/flutter : debugPrintBuildDirtyElements flags to true and looking for a call to scheduleBuildFor for a widget
15:39:00.382: I/flutter : that is labeled "ALREADY IN LIST".
15:39:00.642: I/flutter :
15:39:00.646: I/flutter : When the exception was thrown, this was the stack:
15:39:00.659: I/flutter : #0      BuildOwner.scheduleBuildFor.<anonymous closure> (package:flutter/src/widgets/framework.dart:1522:9)
15:39:00.662: I/flutter : #2      BuildOwner.scheduleBuildFor (package:flutter/src/widgets/framework.dart:1518:12)
15:39:00.665: I/flutter : #3      BuildableElement.markNeedsBuild (package:flutter/src/widgets/framework.dart:2528:11)
15:39:00.668: I/flutter : #4      BuildableElement.dependenciesChanged (package:flutter/src/widgets/framework.dart:2587:5)
15:39:00.671: I/flutter : #5      BuildableElement.activate (package:flutter/src/widgets/framework.dart:2602:7)
15:39:00.673: I/flutter : #6      Element._activateRecursively (package:flutter/src/widgets/framework.dart:2185:13)
15:39:00.676: I/flutter : #7      ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:2693:14)
15:39:00.679: I/flutter : #8      Element._activateRecursively (package:flutter/src/widgets/framework.dart:2187:13)
15:39:00.682: I/flutter : #9      SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:3427:14)
15:39:00.685: I/flutter : #10     Element._activateRecursively (package:flutter/src/widgets/framework.dart:2187:13)
15:39:00.688: I/flutter : #11     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:3427:14)
15:39:00.691: I/flutter : #12     Element._activateRecursively (package:flutter/src/widgets/framework.dart:2187:13)
15:39:00.693: I/flutter : #13     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:3427:14)
15:39:00.696: I/flutter : #14     Element._activateRecursively (package:flutter/src/widgets/framework.dart:2187:13)
15:39:00.699: I/flutter : #15     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:2693:14)
15:39:00.702: I/flutter : #16     Element._activateRecursively (package:flutter/src/widgets/framework.dart:2187:13)
15:39:00.705: I/flutter : #17     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:2693:14)
15:39:00.708: I/flutter : #18     Element._activateRecursively (package:flutter/src/widgets/framework.dart:2187:13)
15:39:00.711: I/flutter : #19     Element._activateWithParent (package:flutter/src/widgets/framework.dart:2178:5)
15:39:00.714: I/flutter : #20     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2122:18)
15:39:00.717: I/flutter : #21     Element.updateChild (package:flutter/src/widgets/framework.dart:1961:12)
15:39:00.720: I/flutter : #22     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3446:14)
15:39:00.723: I/flutter : #23     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.726: I/flutter : #24     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2681:16)
15:39:00.729: I/flutter : #25     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2570:5)
15:39:00.732: I/flutter : #26     ProxyElement.update (package:flutter/src/widgets/framework.dart:2908:5)
15:39:00.734: I/flutter : #27     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.737: I/flutter : #28     RenderObjectElement.updateChildren (package:flutter/src/widgets/framework.dart:3172:26)
15:39:00.740: I/flutter : #29     MultiChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3545:17)
15:39:00.744: I/flutter : #30     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.747: I/flutter : #31     RenderObjectElement.updateChildren (package:flutter/src/widgets/framework.dart:3172:26)
15:39:00.749: I/flutter : #32     MultiChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3545:17)
15:39:00.752: I/flutter : #33     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.754: I/flutter : #34     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3446:14)
15:39:00.757: I/flutter : #35     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.760: I/flutter : #36     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3446:14)
15:39:00.762: I/flutter : #37     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.765: I/flutter : #38     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3446:14)
15:39:00.767: I/flutter : #39     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.771: I/flutter : #40     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3446:14)
15:39:00.773: I/flutter : #41     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.777: I/flutter : #42     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2681:16)
15:39:00.780: I/flutter : #43     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2570:5)
15:39:00.783: I/flutter : #44     StatefulElement.update (package:flutter/src/widgets/framework.dart:2790:5)
15:39:00.785: I/flutter : #45     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.788: I/flutter : #46     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2681:16)
15:39:00.791: I/flutter : #47     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2570:5)
15:39:00.794: I/flutter : #48     StatefulElement.update (package:flutter/src/widgets/framework.dart:2790:5)
15:39:00.796: I/flutter : #49     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.799: I/flutter : #50     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2681:16)
15:39:00.802: I/flutter : #51     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2570:5)
15:39:00.806: I/flutter : #52     StatelessElement.update (package:flutter/src/widgets/framework.dart:2719:5)
15:39:00.809: I/flutter : #53     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.812: I/flutter : #54     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2681:16)
15:39:00.815: I/flutter : #55     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2570:5)
15:39:00.818: I/flutter : #56     StatelessElement.update (package:flutter/src/widgets/framework.dart:2719:5)
15:39:00.821: I/flutter : #57     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.823: I/flutter : #58     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3446:14)
15:39:00.826: I/flutter : #59     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.829: I/flutter : #60     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3446:14)
15:39:00.832: I/flutter : #61     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.835: I/flutter : #62     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2681:16)
15:39:00.838: I/flutter : #63     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2570:5)
15:39:00.841: I/flutter : #64     ProxyElement.update (package:flutter/src/widgets/framework.dart:2908:5)
15:39:00.844: I/flutter : #65     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.846: I/flutter : #66     RenderObjectElement.updateChildren (package:flutter/src/widgets/framework.dart:3172:26)
15:39:00.849: I/flutter : #67     MultiChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3545:17)
15:39:00.852: I/flutter : #68     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.855: I/flutter : #69     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2681:16)
15:39:00.858: I/flutter : #70     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2570:5)
15:39:00.861: I/flutter : #71     StatefulElement.update (package:flutter/src/widgets/framework.dart:2790:5)
15:39:00.864: I/flutter : #72     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.866: I/flutter : #73     RenderObjectElement.updateChildren (package:flutter/src/widgets/framework.dart:3172:26)
15:39:00.869: I/flutter : #74     MultiChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3545:17)
15:39:00.872: I/flutter : #75     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.875: I/flutter : #76     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3446:14)
15:39:00.878: I/flutter : #77     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.880: I/flutter : #78     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3446:14)
15:39:00.883: I/flutter : #79     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.886: I/flutter : #80     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3446:14)
15:39:00.889: I/flutter : #81     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.892: I/flutter : #82     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2681:16)
15:39:00.895: I/flutter : #83     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2570:5)
15:39:00.898: I/flutter : #84     StatelessElement.update (package:flutter/src/widgets/framework.dart:2719:5)
15:39:00.901: I/flutter : #85     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.904: I/flutter : #86     RenderObjectElement.updateChildren (package:flutter/src/widgets/framework.dart:3172:26)
15:39:00.907: I/flutter : #87     MultiChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:3545:17)
15:39:00.910: I/flutter : #88     Element.updateChild (package:flutter/src/widgets/framework.dart:1954:15)
15:39:00.913: I/flutter : #89     _LayoutBuilderElement._layout.<anonymous closure> (package:flutter/src/widgets/layout_builder.dart:120:18)
15:39:00.916: I/flutter : #90     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:1640:19)
15:39:00.919: I/flutter : #91     _LayoutBuilderElement._layout (package:flutter/src/widgets/layout_builder.dart:108:11)
15:39:00.922: I/flutter : #92     RenderObject.invokeLayoutCallback.<anonymous closure> (package:flutter/src/rendering/object.dart:1669:58)
15:39:00.925: I/flutter : #93     PipelineOwner._enableMutationsToDirtySubtrees (package:flutter/src/rendering/object.dart:919:15)
15:39:00.928: I/flutter : #94     RenderObject.invokeLayoutCallback (package:flutter/src/rendering/object.dart:1669:13)
15:39:00.931: I/flutter : #95     _RenderLayoutBuilder.performLayout (package:flutter/src/widgets/layout_builder.dart:207:5)
15:39:00.934: I/flutter : #96     RenderObject.layout (package:flutter/src/rendering/object.dart:1571:21)
15:39:00.936: I/flutter : #97     RenderStack.performLayout (package:flutter/src/rendering/stack.dart:354:15)
15:39:00.939: I/flutter : #98     RenderObject.layout (package:flutter/src/rendering/object.dart:1571:21)
15:39:00.942: I/flutter : #99     _RenderLayoutBuilder.performLayout (package:flutter/src/widgets/layout_builder.dart:209:13)
15:39:00.945: I/flutter : #100    RenderObject.layout (package:flutter/src/rendering/object.dart:1571:21)
15:39:00.948: I/flutter : #101    RenderFlex.performLayout (package:flutter/src/rendering/flex.dart:516:17)
15:39:00.951: I/flutter : #102    RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:1448:21)
15:39:00.953: I/flutter : #103    PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:897:18)
15:39:00.956: I/flutter : #104    BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding.beginFrame (package:flutter/src/rendering/binding.dart:207:19)
15:39:00.959: I/flutter : #105    BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding&WidgetsBinding.beginFrame (package:flutter/src/widgets/binding.dart:297:23)
15:39:00.962: I/flutter : #106    BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:152:5)
15:39:00.965: I/flutter : #107    BindingBase&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:608:15)
15:39:00.968: I/flutter : #108    BindingBase&SchedulerBinding.handleBeginFrame (package:flutter/src/scheduler/binding.dart:543:9)
15:39:00.970: I/flutter : #109    _beginFrame (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:48)
15:39:00.973: I/flutter : (elided one frame from class _AssertionError)
15:39:00.976: I/flutter : ════════════════════════════════════════════════════════════════════════════════════════════════════
```

### LINKED ISSUE #6120: No assert due to duplicate keys
```dart
import 'dart:async';

import 'package:flutter/widgets.dart';

final GlobalKey _kKey = new GlobalKey();
Future main() async {
  runApp(new Block(
      children: new List<Widget>.generate(
          2,
          (_) => new Padding(
              padding: const EdgeInsets.all(8.0),
              child: new Container(
                  key: _kKey,
                  width: 100.0,
                  height: 100.0,
                  decoration: new BoxDecoration(
                      backgroundColor: new Color(0xFFFFFF00)))))));
}

```

### LINKED ISSUE #6121: non-duplicate key assert when duplicate keys
```dart
import 'dart:async';

import 'package:flutter/widgets.dart';

final GlobalKey _kKey = new GlobalKey();
Future main() async {
  runApp(new Block(
      children: new List<Widget>.generate(
          3,
          (_) => new Padding(
              padding: const EdgeInsets.all(8.0),
              child: new Container(
                  key: _kKey,
                  width: 100.0,
                  height: 100.0,
                  decoration: new BoxDecoration(
                      backgroundColor: new Color(0xFFFFFF00)))))));
}

```

Assert:

```
10:35:01.044: I/flutter : ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
10:35:01.055: I/flutter : The following assertion was thrown building RawGestureDetector([GlobalKey 273613476]; state:
10:35:01.058: I/flutter : RawGestureDetectorState(304216675; gestures: vertical drag; behavior: opaque)):
10:35:01.062: I/flutter : scheduleBuildFor() called for a widget for which a build was already scheduled.
10:35:01.065: I/flutter : The method was called for the following element:
10:35:01.068: I/flutter :   DecoratedBox(dirty; renderObject: RenderDecoratedBox NEEDS-LAYOUT DETACHED)
10:35:01.072: I/flutter : The current dirty list consists of:
10:35:01.075: I/flutter :   [DecoratedBox(dirty; renderObject: RenderDecoratedBox NEEDS-LAYOUT DETACHED)]
10:35:01.078: I/flutter : This usually indicates that a widget was rebuilt outside the build phase (thus marking the element
10:35:01.082: I/flutter : as clean even though it is still in the dirty list). This should not be possible and is probably
10:35:01.085: I/flutter : caused by a bug in the widgets framework. Please report it:
10:35:01.087: I/flutter : https://github.com/flutter/flutter/issues/new
10:35:01.090: I/flutter : To debug this issue, consider setting the debugPrintScheduleBuildForStacks and
10:35:01.093: I/flutter : debugPrintBuildDirtyElements flags to true and looking for a call to scheduleBuildFor for a widget
10:35:01.096: I/flutter : that is labeled "ALREADY IN LIST".
10:35:01.243: I/flutter :
10:35:01.247: I/flutter : When the exception was thrown, this was the stack:
10:35:01.272: I/flutter : #0      BuildOwner.scheduleBuildFor.<anonymous closure> (package:flutter/src/widgets/framework.dart:1531:9)
10:35:01.275: I/flutter : #2      BuildOwner.scheduleBuildFor (package:flutter/src/widgets/framework.dart:1527:12)
10:35:01.278: I/flutter : #3      BuildableElement.markNeedsBuild (package:flutter/src/widgets/framework.dart:2538:11)
10:35:01.282: I/flutter : #4      BuildableElement.dependenciesChanged (package:flutter/src/widgets/framework.dart:2584:5)
10:35:01.285: I/flutter : #5      BuildableElement.activate (package:flutter/src/widgets/framework.dart:2599:7)
10:35:01.289: I/flutter : #6      Element._activateRecursively (package:flutter/src/widgets/framework.dart:2194:13)
10:35:01.292: I/flutter : #7      SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:3424:14)
10:35:01.295: I/flutter : #8      Element._activateRecursively (package:flutter/src/widgets/framework.dart:2196:13)
10:35:01.298: I/flutter : #9      ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:2690:14)
10:35:01.301: I/flutter : #10     Element._activateRecursively (package:flutter/src/widgets/framework.dart:2196:13)
10:35:01.305: I/flutter : #11     Element._activateWithParent (package:flutter/src/widgets/framework.dart:2187:5)
10:35:01.307: I/flutter : #12     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2131:18)
10:35:01.310: I/flutter : #13     Element.updateChild (package:flutter/src/widgets/framework.dart:1970:12)
10:35:01.313: I/flutter : #14     SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:3436:14)
10:35:01.316: I/flutter : #15     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2139:14)
10:35:01.319: I/flutter : #16     MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:3532:32)
10:35:01.322: I/flutter : #17     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2139:14)
10:35:01.325: I/flutter : #18     Element.updateChild (package:flutter/src/widgets/framework.dart:1970:12)
10:35:01.327: I/flutter : #19     SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:3436:14)
10:35:01.330: I/flutter : #20     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2139:14)
10:35:01.333: I/flutter : #21     Element.updateChild (package:flutter/src/widgets/framework.dart:1970:12)
10:35:01.336: I/flutter : #22     SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:3436:14)
10:35:01.339: I/flutter : #23     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2139:14)
10:35:01.342: I/flutter : #24     Element.updateChild (package:flutter/src/widgets/framework.dart:1970:12)
10:35:01.346: I/flutter : #25     SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:3436:14)
10:35:01.349: I/flutter : #26     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2139:14)
10:35:01.352: I/flutter : #27     Element.updateChild (package:flutter/src/widgets/framework.dart:1970:12)
10:35:01.355: I/flutter : #28     SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:3436:14)
10:35:01.358: I/flutter : #29     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2139:14)
10:35:01.361: I/flutter : #30     Element.updateChild (package:flutter/src/widgets/framework.dart:1970:12)
10:35:01.364: I/flutter : #31     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2678:16)
10:35:01.367: I/flutter : #32     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2567:5)
10:35:01.370: I/flutter : #33     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:2652:5)
10:35:01.373: I/flutter : #34     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:2768:22)
10:35:01.376: I/flutter : #35     ComponentElement.mount (package:flutter/src/widgets/framework.dart:2647:5)
10:35:01.379: I/flutter : #36     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2139:14)
10:35:01.383: I/flutter : #37     Element.updateChild (package:flutter/src/widgets/framework.dart:1970:12)
10:35:01.386: I/flutter : #38     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2678:16)
10:35:01.389: I/flutter : #39     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2567:5)
10:35:01.392: I/flutter : #40     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:2652:5)
10:35:01.395: I/flutter : #41     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:2768:22)
10:35:01.399: I/flutter : #42     ComponentElement.mount (package:flutter/src/widgets/framework.dart:2647:5)
10:35:01.402: I/flutter : #43     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2139:14)
10:35:01.405: I/flutter : #44     Element.updateChild (package:flutter/src/widgets/framework.dart:1970:12)
10:35:01.409: I/flutter : #45     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2678:16)
10:35:01.412: I/flutter : #46     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2567:5)
10:35:01.415: I/flutter : #47     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:2652:5)
10:35:01.418: I/flutter : #48     ComponentElement.mount (package:flutter/src/widgets/framework.dart:2647:5)
10:35:01.421: I/flutter : #49     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2139:14)
10:35:01.424: I/flutter : #50     Element.updateChild (package:flutter/src/widgets/framework.dart:1970:12)
10:35:01.428: I/flutter : #51     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2678:16)
10:35:01.431: I/flutter : #52     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2567:5)
10:35:01.434: I/flutter : #53     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:2652:5)
10:35:01.437: I/flutter : #54     ComponentElement.mount (package:flutter/src/widgets/framework.dart:2647:5)
10:35:01.440: I/flutter : #55     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2139:14)
10:35:01.444: I/flutter : #56     Element.updateChild (package:flutter/src/widgets/framework.dart:1970:12)
10:35:01.447: I/flutter : #57     RenderObjectToWidgetElement._rebuild (package:flutter/src/widgets/binding.dart:518:16)
10:35:01.449: I/flutter : #58     RenderObjectToWidgetElement.mount (package:flutter/src/widgets/binding.dart:489:5)
10:35:01.453: I/flutter : #59     RenderObjectToWidgetAdapter.attachToRenderTree.<anonymous closure> (package:flutter/src/widgets/binding.dart:435:17)
10:35:01.456: I/flutter : #60     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:1649:19)
10:35:01.459: I/flutter : #61     RenderObjectToWidgetAdapter.attachToRenderTree (package:flutter/src/widgets/binding.dart:434:13)
10:35:01.462: I/flutter : #62     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding&WidgetsBinding.attachRootWidget (package:flutter/src/widgets/binding.dart:336:7)
10:35:01.465: I/flutter : #63     runApp (package:flutter/src/widgets/binding.dart:367:7)
10:35:01.468: I/flutter : #64     main.<main_async_body> (/data/user/0/com.google.fuchsia.armadillo/cache/armadillowvraes/lib/main.dart:11:3)
10:35:01.471: I/flutter : (elided 4 frames from class _AssertionError and package dart:async)
10:35:01.475: I/flutter : ════════════════════════════════════════════════════════════════════════════════════════════════════

```

### PR DESCRIPTION
**Pull request by Hixie**
Fixes https://github.com/flutter/flutter/issues/6109
Fixes https://github.com/flutter/flutter/issues/6120
Fixes https://github.com/flutter/flutter/issues/6121

FYI @apwilson

cc @HansMuller for review

### CODE REVIEW

File: packages/flutter/lib/src/foundation/assertions.dart [hunk 1]
```diff
@@ -198,24 +198,56 @@ class FlutterError extends AssertionError {
...
+          // Regular _AssertionErrors thrown by assert() put the message last, after
+          // some code snippets. This leads to ugly messages. To avoid this, we move
+          // the assertion message up to before the code snippets, separated by a
+          // newline, if we recognise that format is being used.
+          final String message = details.exception.message;
+          final String fullMessage = details.exception.toString();
+          if (message is String && message != fullMessage) {
+            if (fullMessage.length > message.length) {
+              final int position = fullMessage.lastIndexOf(message);
+              if (position == fullMessage.length - message.length &&
```
Comment (HansMuller) [inline packages/flutter/lib/src/foundation/assertions.dart]: The AssertionError toString() method doesn't guarantee formatting. Would it be possible to write a test that made this assumption explicit (and failed if it was violated)?

File: packages/flutter/lib/src/rendering/object.dart [hunk 2]
```diff
@@ -2668,9 +2668,9 @@ abstract class ContainerRenderObjectMixin<ChildType extends RenderObject, Parent
   /// If `after` is null, then this inserts the child at the start of the list,
   /// and the child becomes the new [firstChild].
   void insert(ChildType child, { ChildType after }) {
-    assert(child != this);
-    assert(after != this);
-    assert(child != after);
+    assert(child != this, 'A RenderObject cannot be inserted into itself.');
```
Comment (HansMuller) [inline packages/flutter/lib/src/rendering/object.dart]: NICE

File: packages/flutter/lib/src/widgets/framework.dart [hunk 3]
```diff
@@ -149,31 +149,37 @@ typedef void GlobalKeyRemoveListener(GlobalKey key);
...
   void _register(Element element) {
     assert(() {
       if (_registry.containsKey(this)) {
-        final int oldCount = _debugDuplicates.putIfAbsent(this, () => 1);
-        assert(oldCount >= 1);
-        _debugDuplicates[this] = oldCount + 1;
+        assert(element.widget != null);
+        assert(_registry[this].widget != null);
+        assert(element.widget.runtimeType != _registry[this].widget.runtimeType);
+        _debugIllFatedElements.add(_registry[this]);
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/framework.dart]: Illfated is one word.  If there was something smaller than a nit, this would be it.

Comment (HansMuller) [inline packages/flutter/lib/src/widgets/framework.dart]: I take it back [I looked it up]. It's ill-fated. My mistake.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 4]
```diff
@@ -1846,23 +1899,112 @@ class BuildOwner {
...
   /// After the current call stack unwinds, a microtask that notifies listeners
   /// about changes to global keys will run.
   void finalizeTree() {
     Timeline.startSync('Finalize tree');
     try {
       lockState(() {
-        _inactiveElements._unmountAll();
+        _inactiveElements._unmountAll(); // this unregisters the GlobalKeys
+      });
+      assert(() {
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/framework.dart]: I haven't sorted out all of the logic here.

**Comment by HansMuller** [comment]
LGTM. And thank you for `_debugElementsThatWillNeedToBeRebuiltDueToGlobalKeyShenanigans`

File: packages/flutter/lib/src/foundation/assertions.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/foundation/assertions.dart]: Good call, will do.

Comment (Hixie) [inline packages/flutter/lib/src/foundation/assertions.dart]: done

**Comment by Hixie** [comment]
Added more tests. Will check in on green.

**Pull request merged**